ProxyStarcraft/SpawnLarvaBot.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System.Collections.Generic;
+using ProxyStarcraft.Proto;
+using System.Linq;
+
+namespace ProxyStarcraft
+{
+    /// <summary>
+    /// A bot to handle Queens' <see cref="SpecialAbilityType.SpawnLarva"/> micro.
+    /// </summary>
+    public class SpawnLarvaBot : IBot
+    {
+        public Race Race => Race.Zerg;
+
+        public IReadOnlyList<Command> Act(GameState gameState)
+        {
+            var queens = new List<ZergUnit>();
+            var hatcheries = new List<ZergBuilding>();
+            foreach (var unit in gameState.Units)
+            {
+                if (unit is ZergBuilding zergBuilding && (zergBuilding.ZergBuildingType == ZergBuildingType.Hatchery ||
+                            zergBuilding.ZergBuildingType == ZergBuildingType.Lair ||
+                            zergBuilding.ZergBuildingType == ZergBuildingType.Hive))
+                {
+                    hatcheries.Add(zergBuilding);
+                }
+                
+                else if (unit is ZergUnit zergUnit && zergUnit.ZergUnitType == ZergUnitType.Queen)
+                {
+                    queens.Add(zergUnit);   
+                }
+            }
+            var queensByHatchery = this.ClosestQueenByHatchery(hatcheries, queens);
+            var commands = new List<Command>();
+            foreach (var hatcheryQueenPair in queensByHatchery)
+            {
+                hatcheryQueenPair.Value.SpawnLarva(hatcheryQueenPair.Key, commands);
+            }
+            return commands;
+        }
+
+        private Dictionary<ZergBuilding, ZergUnit> ClosestQueenByHatchery(List<ZergBuilding> hatcheries, List<ZergUnit> queens)
+        {
+            var results = new Dictionary<ZergBuilding, ZergUnit>();
+            if (!queens.Any())
+            {
+                return results;
+            }
+            foreach (var item in hatcheries)
+            {
+                var closestQueen = (ZergUnit)item.GetClosest(queens);
```

Comment:
Because there are no other constraints, this could behave somewhat weirdly in a lot of scenarios. Like if you have two Hatcheries and one Queen, whichever Hatchery is in the list first will 'own' it.

Reply:
Yep! I don't feel bad about that, because spawning larva is about the most important micro that a Zerg player (or bot) has- and this won't assign more than one queen per _built_ and _non-larva-spawning_ hatchery. I should maybe change that to hatcheries that meet the above criteria and aren't the target of any Spawn Larva orders.. or hatcheries that already have more than some number of larva in the row of Locations immediately "south" of them. Or something. That'd give a better sense for when to use Creep Tumors- but then again, that could be in another bot.

Now that you've got me thinking of it though, there might be potentially bad ramifications if one were to, say, expand to a location very far away from any queens. One queen, waddling across un-creeped ground, instead of building another queen at that hatchery. I don't suppose there's any support for "ETA" considering the path a unit would need to take?

Reply:
I see what you're saying about the extra hatcheries issue- I never ran into that, because the Zerg Bot builds a queen from the main hatch whenever a secondary hatchery is going up. I'd seen queens "trading" hatcheries, and figured it didn't matter much.

Having some constraint like "the potential queen must be within 10 units" makes sense.